### PR TITLE
Implement mark order command

### LINKED ISSUE #73: Implement mark order command
To be done after `listOrder` command is implemented and merged.
Command format: `mark ORDER_INDEX`

### PR DESCRIPTION
**Pull request by awdse22**
Close #73 

image

### CODE REVIEW

File: src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public MarkOrderCommand(Index index) {
+        requireNonNull(index);
+        this.index = index;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        Person selectedPerson = model.getSelectedPerson().get();
```
Comment (FooChao) [inline src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:50]: image
I added all these method for my part to follow law of Demeter. I think I will just edit for you later since my part has to come after your part and these method don't exist yet

Comment (FooChao) [reply src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:50]: the method name now is weird now because i cannot refactor now with so many parallel branch around me

File: src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * Creates a marked version of the given {@code Order}.
+     */
+    public static Order createMarkedOrder(Order order) {
+        assert order != null;
+
+        Item item = order.getItem();
+        Date date = order.getDate();
+        Count count = order.getCount();
+        Status status = Status.COMPLETED;
```
Comment (FooChao) [inline src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:75]: Not sure if it is actually important but have u considered the case where the order is already completed

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/DeleteOrderCommandTest.java [hunk 3]
```diff
@@ -38,7 +38,6 @@ public void execute_validIndexOrderList_success() {
         Model expectedModel = model.createCopy();
         Person person = model.getFilteredPersonList().get(3);
         model.updateSelectedPerson(person);
-        expectedModel.updateSelectedPerson(person);
```
Comment (FooChao) [inline src/test/java/seedu/sellsavvy/logic/commands/ordercommands/DeleteOrderCommandTest.java:41]: why did you delete this?

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        Person personToMarkOrderUnder = model.getFilteredPersonList().get(3);
+        model.updateSelectedPerson(personToMarkOrderUnder);
+    }
+
+    @Test
+    public void execute_validIndexOrderList_success() {
+        MarkOrderCommand markOrderCommand = new MarkOrderCommand(INDEX_FIRST_ORDER);
+
+        Model expectedModel = model.createCopy();
+        Order orderToBeMarked = expectedModel.getFilteredPersonList().get(3)
```
Comment (FooChao) [inline src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:39]: maybe you could standardise and put INDEX_FOUNRTH_PERSON.getZeroBased?

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                .getOrderList().get(INDEX_FIRST_ORDER.getZeroBased());
+        Order markedOrder = MarkOrderCommand.createMarkedOrder(orderToBeMarked);
+
+        expectedModel.getFilteredPersonList().get(3).getOrderList().setOrder(orderToBeMarked, markedOrder);
```
Comment (FooChao) [inline src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:43]: same as above

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        String expectedMessage = String.format(MarkOrderCommand.MESSAGE_MARK_ORDER_SUCCESS,
+                Messages.format(markedOrder));
+
+        assertCommandSuccess(markOrderCommand, model, expectedMessage, expectedModel);
+    }
+
+    @Test
+    public void execute_invalidIndexOrderList_throwsCommandException() {
+        Person person = model.getFilteredPersonList().get(3);
```
Comment (FooChao) [inline src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:52]: same as above

File: src/test/java/seedu/sellsavvy/storage/JsonAdaptedOrderTest.java [hunk 7]
```diff
@@ -11,15 +11,18 @@
...
 public class JsonAdaptedOrderTest {
     private static final String INVALID_ITEM = " ";
     private static final String INVALID_COUNT = "0";
     private static final String INVALID_DATE = "2-2-2";
+    private static final String INVALID_STATUS = "not pending";
 
     private static final String VALID_ITEM = BLOCKS.getItem().toString();
     private static final String VALID_COUNT = BLOCKS.getCount().toString();
     private static final String VALID_DATE = BLOCKS.getDate().toString();
+    private static final String VALID_STATUS = BLOCKS.getStatus().toString();
```
Comment (FooChao) [inline src/test/java/seedu/sellsavvy/storage/JsonAdaptedOrderTest.java:25]: Is there really no success test cases or you forgot to edit?

**Review by FooChao — COMMENTED** [review]
LGTM other than some small nitpicks.

File: docs/UserGuide.md [hunk 8]
```diff
@@ -185,12 +185,27 @@ Format: `deleteOrder ORDER_INDEX`
...
-* The index and quantity **must be a positive integer** 1, 2, 3, …​
+* The index **must be a positive integer** 1, 2, 3, …​
 
 Examples:
 * `deleteOrder 1` deletes the order with index 1 from the selected person.
 * `listOrder 1` followed by `deleteOrder 2` selects the 1st person in the address book and deletes the 2nd order under the 1st person.
 
+### Mark an order as complete : `markOrder`
+
+Marks an order under the person whose orders are displayed as complete.
```
Comment (FooChao) [inline docs/UserGuide.md:196]: isn't is completed?

File: docs/UserGuide.md [hunk 9]
```diff
@@ -185,12 +185,27 @@ Format: `deleteOrder ORDER_INDEX`
...
+
+Format: `markOrder ORDER_INDEX`
+
+* A person's order list must be selected using `listOrder` before marking an order from that person.
+* Marks an order under the selected person at the specified `ORDER_INDEX` as complete.
```
Comment (FooChao) [inline docs/UserGuide.md:201]: isn't is completed?

File: docs/UserGuide.md [hunk 10]
```diff
@@ -185,12 +185,27 @@ Format: `deleteOrder ORDER_INDEX`
...
+* The order index refers to the index number shown in the **displayed order list**.
+* The index **must be a positive integer** 1, 2, 3, …​
+
+Examples:
+* `markOrder 1` marks the order with index 1 from the displayed order list as complete.
```
Comment (FooChao) [inline docs/UserGuide.md:206]: ^

File: docs/UserGuide.md [hunk 11]
```diff
@@ -185,12 +185,27 @@ Format: `deleteOrder ORDER_INDEX`
...
+* `listOrder 1` followed by `markOrder 2` selects the 1st person in the address book and marks the 2nd order under the 1st person as complete.
```
Comment (FooChao) [inline docs/UserGuide.md:207]: ^

File: docs/UserGuide.md [hunk 12]
```diff
@@ -242,13 +257,14 @@ _Details coming soon ..._
...
+**Add a person**                | `add n/NAME p/PHONE_NUMBER e/EMAIL a/ADDRESS [t/TAG]…​` <br> e.g., `add n/James Ho p/22224444 e/jamesho@example.com a/123, Clementi Rd, 1234665 t/friend t/colleague`
+**Clear**                       | `clear`
+**Delete a person**             | `delete INDEX`<br> e.g., `delete 3`
+**Edit a person**               | `edit INDEX [n/NAME] [p/PHONE_NUMBER] [e/EMAIL] [a/ADDRESS] [t/TAG]…​`<br> e.g.,`edit 2 n/James Lee e/jameslee@example.com`
+**Find person(s)**              | `find KEYWORD [MORE_KEYWORDS]`<br> e.g., `find James Jake`
+**List all persons**            | `list`
+**Add an order**                | `addOrder INDEX i/ITEM d/DATE [c/QUANTITY]` <br> e.g., `addOrder 2 i/Lamp d/20-11-2024 c/3`
+**List all orders**             | `listOrder INDEX`<br> e.g., `listOrder 3`
+**Delete an order**             | `deleteOrder ORDER_INDEX`<br> e.g., `deleteOrder 2`
+**Mark an order as complete**   | `markOrder ORDER_INDEX`<br> e.g., `markOrder 2`
```
Comment (FooChao) [inline docs/UserGuide.md:269]: ^

**Review by FooChao — COMMENTED** [review]
Based on your javadocs it is completed but user guide says complete.

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/DeleteOrderCommandTest.java (hunk 3 above)
Comment (awdse22) [reply src/test/java/seedu/sellsavvy/logic/commands/ordercommands/DeleteOrderCommandTest.java:41]: this line is completely redundant when I was referring to DeleteOrderCommandTest, felt it wasn't exactly worth creating another PR/issue for.

File: src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java (hunk 1 above)
Comment (awdse22) [reply src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:50]: alright, noted.

File: src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java (hunk 2 above)
Comment (awdse22) [reply src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:75]: have considered but ultimately the result is the same, it's like editing a Person with name "Alice" to "Alice". Would you have preferred if a message is displayed saying "The order has already been marked as completed"?

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java (hunk 4 above)
Comment (awdse22) [reply src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:39]: will change by creating a new integer constant for `MarkOrderCommandTest` called `TARGET_INDEX`

File: src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java (hunk 2 above)
Comment (FooChao) [reply src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:75]: I think that would been better because if a user mark something that is already completed it is likely that it is a typo plus it may cause unnecessary re-rendering of the gui if we replace stuff unnecessarily in the orderList.

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java (hunk 4 above)
Comment (FooChao) [reply src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:39]: shouldn't it be in TypicalIndex? In fact Sutong already created it in previous commit

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java (hunk 5 above)
Comment (awdse22) [reply src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:43]: changed

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java (hunk 6 above)
Comment (awdse22) [reply src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:52]: changed

File: src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java (hunk 4 above)
Comment (awdse22) [reply src/test/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommandTest.java:39]: oh my bad, changing to `INDEX_FOURTH_PERSON.getZeroBased()`

File: docs/UserGuide.md (hunk 8 above)
Comment (awdse22) [reply docs/UserGuide.md:196]: will change to completed

File: docs/UserGuide.md (hunk 10 above)
Comment (awdse22) [reply docs/UserGuide.md:206]: will change

File: docs/UserGuide.md (hunk 9 above)
Comment (awdse22) [reply docs/UserGuide.md:201]: will change

File: docs/UserGuide.md (hunk 11 above)
Comment (awdse22) [reply docs/UserGuide.md:207]: will change

File: docs/UserGuide.md (hunk 12 above)
Comment (awdse22) [reply docs/UserGuide.md:269]: will change

File: src/test/java/seedu/sellsavvy/storage/JsonAdaptedOrderTest.java (hunk 7 above)
Comment (awdse22) [reply src/test/java/seedu/sellsavvy/storage/JsonAdaptedOrderTest.java:25]: image
this is the success test case which has the original valid values of `BLOCKS`

File: src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java (hunk 2 above)
Comment (awdse22) [reply src/main/java/seedu/sellsavvy/logic/commands/ordercommands/MarkOrderCommand.java:75]: will create a new check under the `MarkOrderCommand::execute` method which throws a `CommandException` if the order is already marked as completed.

**Review by FooChao — APPROVED** [review]
LGTM but let them see first especially @yu-sutong he need to understand how it works

**Review by yu-sutong — APPROVED** [review]
Looks really good to me.
